jobbergate-api/jobbergate_api/tests/apps/job_scripts/test_job_script_files.py

```diff
@@ -0,0 +1,275 @@
+        supporting_file_content = "test-support-file"
+        supporting_file_path = work_dir / JOBSCRIPTS_SUPPORTING_FILES_FOLDER / supporting_file_name
+        supporting_file_path.parent.mkdir(parents=True, exist_ok=True)
+        supporting_file_path.write_text(supporting_file_content)
+
+        assert main_file_path.is_file()
+        assert supporting_file_path.is_file()
+
+        JobScriptFiles.delete_from_s3(job_script_id)
+
+        assert not main_file_path.is_file()
+        assert not supporting_file_path.is_file()
+
+    def test_write_to_s3(self, mocked_file_manager_factory):
+        """
+        Test that JobScriptFiles are written as expected.
+        """
+        job_script_id = 1
+
+        work_dir = mocked_file_manager_factory / JOBSCRIPTS_WORK_DIR / str(job_script_id)
+        output_dir = Path("test-output-dir")
+
+        main_file_name = output_dir / "jobbergate.sh"
+        main_file_content = "test-main-file"
+        main_file_path = work_dir / JOBSCRIPTS_MAIN_FILE_FOLDER / main_file_name
+
+        supporting_file_name = output_dir / "support.sh"
+        supporting_file_content = "test-support-file"
+        supporting_file_path = work_dir / JOBSCRIPTS_SUPPORTING_FILES_FOLDER / supporting_file_name
+
+        JobScriptFiles(
+            main_file_path=main_file_name,
+            files={
+                main_file_name: main_file_content,
+                supporting_file_name: supporting_file_content,
+            },
+        ).write_to_s3(job_script_id)
+
+        assert main_file_path.is_file()
+        assert main_file_path.read_text() == main_file_content
+
+        assert supporting_file_path.is_file()
+        assert supporting_file_path.read_text() == supporting_file_content
+
+    def test_io_integration_with_s3(self, mocked_file_manager_factory):
+        """
+        End-to-end test where the JobScriptFiles are written and read from s3.
+        """
+        job_script_id = 1
+        output_dir = Path("test-output-dir")
+
+        main_file_name = output_dir / "jobbergate.sh"
+        main_file_content = "test-main-file"
+
+        supporting_file_name = output_dir / "support.sh"
+        supporting_file_content = "test-support-file"
+
+        actual_jobscript_files = JobScriptFiles(
+            main_file_path=main_file_name,
+            files={
+                main_file_name: main_file_content,
+                supporting_file_name: supporting_file_content,
+            },
+        )
+
+        actual_jobscript_files.write_to_s3(job_script_id)
+
+        desired_jobscript_files = JobScriptFiles.get_from_s3(job_script_id)
+
+        assert desired_jobscript_files == actual_jobscript_files
+
+        JobScriptFiles.delete_from_s3(job_script_id)
+
+        with pytest.raises(
+            ValueError,
+            match="One main file is expected for a job-script, found 0",
+        ):
+            JobScriptFiles.get_from_s3(job_script_id)
+
+    def test_render_from_application__sucess(
```

Comment:
You need one unit test (at least) for each failure condition that can happen in the `render_from_application` method.

To reduce the difficulty of this and to improve the readability of that method, it would probably be good to break it up into multiple "private" methods that are called in succession.